docs/guides/spmd_api.md

```diff
@@ -0,0 +1,429 @@
+
+You can also check the underlying `.sharding` of the JAX array, which gives a hint on the way it is partitioned.
+<!-- #endregion -->
+
+```python id="UEj1A2BhhUcC"
+print(type(initialized_state.params['ScanDotReluDot_0']['W1']))
+print(type(initialized_state.params['ScanDotReluDot_0']['W1'].value))
+print(initialized_state.params['ScanDotReluDot_0']['W1'].value.shape)
+```
+
+```python id="-Hhp_glBOL1R"
+print(initialized_state.params['ScanDotReluDot_0']['W1'].value.sharding)
+```
+
+<!-- #region id="Yu7ohrBzN-eP" -->
+You can use `jax.tree_map` to perform mass computation on a dict of boxed params, in the same way as on a dict of JAX arrays.
+<!-- #endregion -->
+
+```python id="aqwt7F0Kmn0f"
+diff = jax.tree_map(
+    lambda a, b: a - b, 
+    initialized_state.params['ScanDotReluDot_0'], initialized_state.params['ScanDotReluDot_0'])
+print(jax.tree_map(jnp.shape, diff))
+diff_array = diff['W1'].unbox()
+print(type(diff_array))
+print(diff_array.shape)
+```
+
+<!-- #region id="dcLFwH1dhr3v" -->
+## `pjit` the train step and inference 
+
+Now we can do the same and make a `pjit`ted training step.
+<!-- #endregion -->
+
+```python id="KF5lmJ5FwxqJ"
+def train_step(state, x):
+  # fake loss fn
+  def loss_unrolled(params):
+    y = model.apply({'params': params}, x)
+    return y.sum()
+  grad_fn = jax.grad(loss_unrolled)
+  grads = grad_fn(state.params)
+  state = state.apply_gradients(grads=grads)
+  return state
+
+pjit_step_fn = pjit(train_step,
+                    (state_spec, x_spec),  # input annotations
+                    state_spec,            # output annotations
+                    )
+```
+
+```python id="F5gy_gb_mOqR"
+with mesh:
+  new_state = pjit_step_fn(initialized_state, x)
+```
+
+```python id="0BE_qTV-vbfo"
+def apply(state, x):
+  return state.apply_fn({'params': state.params}, x)
+
+pjit_apply_fn = pjit(apply, (state_spec, x_spec), x_spec)
+
+with mesh:
+  y = pjit_apply_fn(new_state, x)
+print(type(y))
+print(y.dtype)
+print(y.shape)
+```
+
+<!-- #region id="vXY2mx7WjOyv" -->
+## Profiling
+
+If you are running on a TPU pod or pod slice, you can use the `block_all` util below to profile the cell below to observe a performance increase.
+<!-- #endregion -->
+
+```python id="A2Jn7kktw_UZ"
+%%timeit
+
+# check this out by profiling, your model is made model parallel automatically by XLA:SPMD!
+def block_all(xs):
+  jax.tree_map(lambda x: x.block_until_ready(), xs)
+  return xs
+
+with mesh:
+  new_state = block_all(pjit_step_fn(initialized_state, x))
+```
+
+<!-- #region id="GUXkJiQkotfm" -->
+# Logical axis annotation
+
+JAX auto SPMD encourages user to explore different sharding layouts to find the optimal one. To this end, in Flax you actually can annotate the dimensions of any array with more descriptive axis names, instead of only the device mesh axis names (i.e., `data` and `model`). 
```

Comment:
```suggestion
JAX auto SPMD encourages users to explore different sharding layouts to find the optimal one. To this end, in Flax you actually can annotate the dimensions of any array with more descriptive axis names, instead of only the device mesh axis names (i.e., `data` and `model`). 
```